pyg_lib/csrc/sampler/cuda/random_walk_kernel.cu

```diff
@@ -0,0 +1,83 @@
+#include <ATen/ATen.h>
+#include <ATen/cuda/CUDAContext.h>
+#include <torch/library.h>
+
+namespace pyg {
+namespace sampler {
+
+namespace {
+
+#define THREADS 1024
+#define BLOCKS(N) (N + THREADS - 1) / THREADS
+
+#define CUDA_1D_KERNEL_LOOP(scalar_t, i, n)                       \
+  for (scalar_t i = blockIdx.x * blockDim.x + threadIdx.x; i < n; \
+       i += blockDim.x * gridDim.x)
+
+template <typename scalar_t>
+__global__ void random_walk_kernel_impl(
+    const scalar_t* __restrict__ rowptr_data,
+    const scalar_t* __restrict__ col_data,
+    const scalar_t* __restrict__ seed_data,
+    const float* __restrict__ rand_data,
+    scalar_t* __restrict__ out_data,
+    int64_t num_seeds,
+    int64_t walk_length) {
+  CUDA_1D_KERNEL_LOOP(scalar_t, i, num_seeds) {
+    auto v = seed_data[i];
+    out_data[i] = v;
+
+    for (scalar_t j = 0; j < walk_length; ++j) {
+      auto row_start = rowptr_data[v], row_end = rowptr_data[v + 1];
+      if (row_end - row_start > 0) {
+        auto rand = rand_data[j * num_seeds + i];
+        v = col_data[row_start + scalar_t(rand * (row_end - row_start))];
+      }
+      // For isolated nodes, this will add a fake self-loop.
+      // This does not do any harm when used in within a `node2vec` model.
+      out_data[(j + 1) * num_seeds + i] = v;
+    }
+  }
+}
+
+at::Tensor random_walk_kernel(const at::Tensor& rowptr,
+                              const at::Tensor& col,
+                              const at::Tensor& seed,
+                              int64_t walk_length,
+                              double p,
+                              double q) {
+  TORCH_CHECK(rowptr.is_cuda(), "'rowptr' must be a CUDA tensor");
```

Comment:
I'm curious that: will device information be checked when torch dispatch CPU/GPU kernels?